Add monitoring section

GiovanniGiacometti opened the pull request:
Ho messo i nuovi 4 file creati in una nuova sezione della user guide. Poi valuteremo come inserirli una volta terminati anche gli altri capitoli.

Sarei contento se anche @andreacerasani, @alesnow97, @lambertods volessero dare una letta. Sentitevi liberi di farlo più avanti, eventualmente anche a PR conclusa. Sicuramente un vostro parere sarebbe molto utile dato che siete  un po' meno dentro le logiche platform. Grazie!

GiovanniGiacometti committed:
Start monitoring chapter

GiovanniGiacometti committed:
Why we need monitoring, start target and metrics

GiovanniGiacometti committed:
Monitoring targets

GiovanniGiacometti committed:
Finished base monitoring page + detection event and explainability

detection event rule page is already there

GiovanniGiacometti committed:
Add metric description

GiovanniGiacometti committed:
Rearrange docs

GiovanniGiacometti committed:
Complete detection event

GiovanniGiacometti committed:
Rename monitoring.md to index.md, drift explainability report

GiovanniGiacometti committed:
Modification

GiovanniGiacometti committed:
Merge remote-tracking branch 'origin/dev' into dev-monitoring

GiovanniGiacometti committed:
Fix links and small impr

GiovanniGiacometti committed:
Add customer id in detection event page (will be added soon)

alelavml3 commented:
Va bene, la leggo mentre andrò al gabinetto come lettura di passatempo. Intanto ne approfitto e rispondo con una PR sulla documentazione Task che ho terminato ieri.

GiovanniGiacometti committed:
Add section in monitoring index.md to explain how to access drift status

GiovanniGiacometti commented:
> mentre andrò al gabinetto come lettura di passatempo

Sono contento di stimolare la tua attività

alelavml3 commented on md-docs/user_guide/monitoring/index.md:
```diff
@@ -0,0 +1,122 @@
+Monitoring, also known as __Drift Detection__ in the literature, refers the process of continuously tracking the performance of a model 
+and the distribution of the data it is operating on. 
+
+## How does the MLCube Platform perform Monitoring?
```
ML cube al posto di MLCube. Anche un po' in tutto il documento

alelavml3 replied:
In generale, non so se parlare direttamente attraverso i metodi del client sdk oppure in modo più astratto dicendo che bisogna impostare una reference e che poi si inviano i dati di produzione che vengono confrontati dai diversi algoritmi rispetto alla reference. In questo modo restiamo più indipendenti a futuri cambiamenti e ci focalizziamo maggiormente sui concetti.

alelavml3 commented on md-docs/user_guide/monitoring/index.md:
```diff
@@ -0,0 +1,122 @@
+is drifting away from the expected behavior and that corrective actions should be taken.
+
+In more practical terms, the [set_model_reference] method can be used to specify the time period where the reference of a given model should be placed. As a consequence,
+all algorithms associated with the specified model (not just those monitoring the performance, but also those operating on the data used by the model) will
```
Non è specificato che per "all algorithms" ti stai riferendo ai monitoring algorithm perché nel paragrafo prima li hai chiamati "statistical techniques". Credo basti aggiungere una frase che li introduce dicendo che le diverse statistical techniques prendono il nome di monitoring algorithms e che ce ne sono diversi a seconda dell'entità che si sta monitorando.

alelavml3 commented on md-docs/user_guide/monitoring/index.md:
```diff
@@ -0,0 +1,122 @@
+[add_historical_data] method.
+
+After setting the reference, the [add_production_data] method can be used to send production data to the platform. This data will be analyzed by the monitoring algorithms
+and, if a significant difference is detected, an alarm will be raised, in the form of a [DetectionEvent]. 
```
metterei lo spazio a Detection Event come hai fatto successivamente

alelavml3 commented on md-docs/user_guide/monitoring/index.md:
```diff
@@ -0,0 +1,122 @@
+### Targets and Metrics
+
+After explaining why monitoring is so important in modern AI systems and detailing how it is performed in the ML cube Platform, 
+we can introduce the concepts of Monitoring Targets and Monitoring Metrics. They both represent quantities that the MLCube Platform monitors, but they differ in their nature.
```
aggiungerei un breve paragrafo che spiega perché ML cube Platform ha scelto di monitorare tante cose e non ad esempio solo l'errore. Risposta: per fornire una visione complessiva di come l'ambiente di produzione evolve nel tempo e per fornire più informazioni possibile per comprendere i drift globali.

alelavml3 replied:
Aggiungerei che i monitoring target e metrics sono definiti automaticamente da ML cube Platform a partire dal Task Type e dalla Data Structure

alelavml3 commented on md-docs/user_guide/monitoring/index.md:
```diff
@@ -0,0 +1,122 @@
+Notice that while some targets were specifically designed for a certain task type, others are more general and can be used in different contexts. 
+Nonetheless, the platform might not support yet all possible combinations. The table will be updated as new targets are added to the product.
+
+| **Monitoring Target**          |   **REGRESSION**   | **CLASSIFICATION_BINARY** | **CLASSIFICATION_MULTICLASS** | **CLASSIFICATION_MULTILABEL** | **OBJECT_DETECTION** |      **RAG**       |
```
Io ho usato un altro stile e altre icone. Direi che basta uniformarci alla fine

alelavml3 commented on md-docs/user_guide/monitoring/index.md:
```diff
@@ -0,0 +1,122 @@
+#### Monitoring Metrics
+
+A Monitoring Metric is a generic quantity that can be computed on a Monitoring Target. They enable the monitoring of specific
+aspects of a target, which might help in identifying the root cause of a drift, as well as defining the corrective actions to be taken.
```
aspects of an entity

alelavml3 commented on md-docs/user_guide/monitoring/index.md:
```diff
@@ -0,0 +1,122 @@
+A Monitoring Metric is a generic quantity that can be computed on a Monitoring Target. They enable the monitoring of specific
+aspects of a target, which might help in identifying the root cause of a drift, as well as defining the corrective actions to be taken.
+
+The following table display the monitoring metrics supported, along with their monitoring target and the conditions
```
the following table displays

alelavml3 commented on md-docs/user_guide/monitoring/index.md:
```diff
@@ -0,0 +1,122 @@
+The following table display the monitoring metrics supported, along with their monitoring target and the conditions
+under which they are actually monitored. Notice that also this table is subject to changes, as new metrics will be added.
+
+| **Monitoring Metric** | Description                                              |              **Monitoring Target**               |             **Conditions**             |
```
Aggiungerei le classi per toxicity, emotion e sentiment e una spiegazione in più su cosa si intende per uncertainty

GiovanniGiacometti replied:
Le classi di emotion sono una quindicina, dici di metterle sempre nella tabella?

alelavml3 replied:
In analytics le mostreremo? Comunque direi di sì anche se sono tante

GiovanniGiacometti replied:
Eh sì lì si vedono. Va bene, allora le inserisco

alelavml3 commented on md-docs/user_guide/monitoring/index.md:
```diff
@@ -0,0 +1,122 @@
+
+### Monitoring Status
+
+All the entities being monitored are associated with a status, which is defined according to the enumeration [MonitoringStatus]. 
```
Anche qui resterei indipendente dall'SDK perché se uno lavora in web app gli basta e avere i tre valori. Enumeration è un po' troppo specifica/tecnica come parola.

alelavml3 commented on md-docs/user_guide/monitoring/index.md:
```diff
@@ -0,0 +1,122 @@
+
+All the entities being monitored are associated with a status, which is defined according to the enumeration [MonitoringStatus]. 
+
+The status can be one of the following:
```
Sarebbe molto bello creare un diagramma di flusso con una mini spiegazione di quando e come avviene il cambio di stato (anche del fatto che si possa tornare a OK dopo un Drift che magari non è intuitivo).

Puoi anche creare un'immagine su power point, non è necessario fare il flusso da qua.

Link della guida:
- https://mermaid.js.org/syntax/flowchart.html
- https://squidfunk.github.io/mkdocs-material/reference/diagrams/

Esempietto:
```
  ``` mermaid
  graph LR
    A[Ok] --> B[Warning];
    B --> A;
    A --> C[Drift];
    C --> B;
    B --> C;
    A --> B;
 \```
```

alelavml3 commented on md-docs/user_guide/monitoring/index.md:
```diff
@@ -0,0 +1,122 @@
+You can explore more about detection events and how you can set up automatic actions upon their reception in the [Detection Event] 
+and the [Detection Event Rule] sections respectively.
+
+### Targets and Metrics
```
Non sarebbe male aggiungere un'immagine come questa immagine che puoi trovare nella presentazione platform overview
![immagine](https://github.com/user-attachments/assets/842a818d-0a85-462f-8628-c4ca5f3a7c8c)

alelavml3 commented on md-docs/user_guide/monitoring/detection_event.md:
```diff
@@ -0,0 +1,47 @@
+# Detection Event
```
MLCube Platform al posto di ML cube un po' nella pagina

alelavml3 commented on md-docs/user_guide/monitoring/detection_event.md:
```diff
@@ -0,0 +1,47 @@
+
+A [Detection Event] is raised by the MLCube Platform when a significant change is detected in one of the entities being monitored.
+
+An event is characterized by the following attributes:
```
Come scelta di stile, ho preferito non parlare degli attributi in "modalità python" ma in generale sempre riferendomi ai concetti (puoi trovare gli esempi nella pagina del task o del data schema). La preferisco perché rimaniamo indipendenti dall'sdk. Comunque se vuoi ne possiamo parlare assieme.

Ad esempio, per event type farei così:

```
- Event Type: what the event actually is, possible values are:
    - Warning On: when the monitored entity has a suspicious slight change of distribution that could turn into drift
    - ...
```

alelavml3 commented on md-docs/user_guide/monitoring/detection_event.md:
```diff
@@ -0,0 +1,47 @@
+the criticality of the detected drift.
+- `monitoring_target`: the [MonitoringTarget] being monitored.
+- `monitoring_metric`: the [MonitoringMetric] that triggered the event, if the event is related to a metric.
+- `model_name`: the name of the model that raised the event.
```
aggiungerei che è presente solo quando il monitoring target riguarda il modello

alelavml3 commented on md-docs/user_guide/monitoring/detection_event.md:
```diff
@@ -0,0 +1,47 @@
+You can access the detection events generated by the Platform in two ways:
+
+- **SDK**: the [get_detection_events] method can be used to retrieve all detection events for a specific task programmatically.
+- **WebApp**: navigate to the **`Detection `** section located in the task page's sidebar. Here, all detection events are displayed in a table, 
```
Aggiungerei che si possono vedere anche dalla Task home page (io l'ho già messo nella mia pagina di doc ma lo aggiungerei anche qua)

alelavml3 commented on md-docs/user_guide/monitoring/detection_event_rules.md:
```diff
@@ -0,0 +1,61 @@
+# Detection Event Rules
```
Sempre MLCube Platform in giro

alelavml3 commented on md-docs/user_guide/monitoring/detection_event_rules.md:
```diff
@@ -0,0 +1,61 @@
+When a detection event is produced, the MLCube Platform reviews all the detection event rules you have set 
+and triggers those matching the event.
+
+Rules are specific to a task and require the following parameters:
```
Stesso ragionamento nello spiegare gli attributi meno come "codice" ma più come concetto

alelavml3 commented on md-docs/user_guide/monitoring/detection_event_rules.md:
```diff
@@ -0,0 +1,61 @@
+- `name`: a descriptive label of the rule.
+- `task_id`: the unique identifier of the task to which the rule belongs.
+- `detection_event_type`: the [DetectionEventType] that the rule matches.
+- `severity`: the [DetectionEventSeverity] of the event.
```
Aggiungere che se non viene specificata allora si intente qualsiasi severity.
Mancano i seguenti attributi che dipendono dal monitoring target selezionato:
- monitoring metric
- model

PS: espanderei questa cosa anche ai event type e ai targets con anche cose del tipo "Tutti", "Tutti tranne A, B, C" eccetera. Se sei d'accordo possiamo creare una issue

GiovanniGiacometti replied:
Model non sarebbe model_name? Se intendi la specification, non c'è, ho già creato una issue apposta perché nell'sdk non esiste mai la specification

GiovanniGiacometti replied:
Per l'altra cosa sì, ci sta, potrebbe essere carino, ma lo mettere in una nuova issue

GiovanniGiacometti replied:
Issue creata

alelavml3 commented on md-docs/user_guide/monitoring/detection_event_rules.md:
```diff
@@ -0,0 +1,61 @@
+- `actions`: A sequential list of actions to be executed when the rule is triggered.
+
+## Detection Event Actions
+Two types of actions are currently supported: notification and retrain.
```
Ci sono anche le action di creazione della plot configuration

GiovanniGiacometti replied:
Non c'è questa opzione nell'sdk..

alelavml3 replied:
grave mancanza allora. C'è sul web

alelavml3 commented on md-docs/user_guide/monitoring/drift_explainability.md:
```diff
@@ -0,0 +1,37 @@
+# Drift Explainability
```
Povera __ML cube__ Platform

alelavml3 commented on md-docs/user_guide/monitoring/drift_explainability.md:
```diff
@@ -0,0 +1,37 @@
+## Structure
+
+A Drift Explainability Report consists in comparing the reference data and the portion of production data where the drift was identified, hence 
+those belonging to the new concept. Notice that these reports are generated after a sufficient amount of samples has been collected after the drift.
```
forse è la prima volta che si parla di _concept_ quindi o usi "data distribution" oppure spieghi che cos'è

alelavml3 commented on md-docs/user_guide/monitoring/drift_explainability.md:
```diff
@@ -0,0 +1,37 @@
+
+A Drift Explainability Report consists in comparing the reference data and the portion of production data where the drift was identified, hence 
+those belonging to the new concept. Notice that these reports are generated after a sufficient amount of samples has been collected after the drift.
+If the distribution moves back to the reference before enough samples are collected, the report might not be generated.
```
That's why elements componing the report needs a given amount of data to guarantee statistical reliability of the shown results. (sicuramente la si può scrivere meglio)

alelavml3 commented on md-docs/user_guide/monitoring/drift_explainability.md:
```diff
@@ -0,0 +1,37 @@
+Observed and analyzed together, they should provide a comprehensive understanding of the drift and its underlying causes.
+These are the entities currently available:
+
+- `Feature Importance`: it's a barplot that illustrates how the significance of each feature differs between the reference 
```
Sarebbe bello inserire un'immagine di esempio a cui fai riferimento nella spiegazione. Uguale anche per il punto sotto.
Intanto che ci siamo vuoi anche aggiungere la drift score?

GiovanniGiacometti replied:
Ma intendi tipo i plot ottenuti con il demo?

alelavml3 replied:
Sì, esatto

alelavml3 reviewed (changes requested):
Tendenzialmente va bene. Punti principali:
- ML cube e non MLCube
- Spiegazioni meno con stile tecnico tipo Python ma più concettuali (ho messo un po' di commenti ma possiamo parlarne a voce che magari mi spiego meglio)
- Qualche immagine o grafico in più a supporto delle sipiegazioni

GiovanniGiacometti committed:
Merge remote-tracking branch 'origin/dev' into dev-monitoring

GiovanniGiacometti committed:
Quick fixes post pr

GiovanniGiacometti committed:
Second set of corrections post pr

Plot configurations, more general description of monitoring

GiovanniGiacometti committed:
Third set of corrections post pr

General Detection Event description, typo fix, possible metric, values, state diagram for monitoring status

GiovanniGiacometti committed:
Detection event rules page

GiovanniGiacometti committed:
Merge remote-tracking branch 'origin/dev' into dev-monitoring

GiovanniGiacometti committed:
Adapted new style; small rewrites

GiovanniGiacometti committed:
Drift Explainability Report Images

GiovanniGiacometti committed:
Monitoring overview image

GiovanniGiacometti committed:
Last minor fixes

GiovanniGiacometti committed:
New monitoring status state diagram image

GiovanniGiacometti committed:
New state diagram for monitoring status
